### PR TITLE
Add client for interacting with Bitbucket DC's upcoming deployment status API

### PR DESCRIPTION
**Pull request by khughes-atlassian**
Bitbucket Server is looking into showing deployment status on pull requests and commits. This is the first PR to the feature branch `feature/deployment-notifications` with the client code to:
* Create a deployment status object that aligns with Bitbucket's unreleased REST API
* Send that object to Bitbucket Server

Future PRs will handle recognising when a deployment has happened. They will most likely be in the form of a `Notifier` step that consumers can use for freestyle jobs and a `BuildWrapper` that consumers can use to wrap the deployment steps on a pipelines job

### CODE REVIEW

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/model/deployment/BitbucketCDCapabilities.java
Comment (EliSalter-atl) [inline src/main/java/com/atlassian/bitbucket/jenkins/internal/model/deployment/BitbucketCDCapabilities.java:14]: just checking that all is okay because we have

`public static final String DEPLOYMENTS_CAPABILITY_KEY = "deployment";` (no s on the end) inside `AtlassianServerCapabilities`.

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/model/deployment/BitbucketCDCapabilities.java
Comment (EliSalter-atl) [inline src/main/java/com/atlassian/bitbucket/jenkins/internal/model/deployment/BitbucketCDCapabilities.java:20]:
```suggestion
        this.cdCapabilities = unmodifiableSet(requireNonNull(cdCapabilities, "Deployment capability missing."));
```

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/model/deployment/BitbucketCDCapabilities.java
Comment (EliSalter-atl) [reply src/main/java/com/atlassian/bitbucket/jenkins/internal/model/deployment/BitbucketCDCapabilities.java:14]:
```suggestion
    public static final String DEPLOYMENTS_CAPABILITY = "deployments";
```

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/model/deployment/BitbucketDeploymentEnvironment.java
Comment (EliSalter-atl) [inline src/main/java/com/atlassian/bitbucket/jenkins/internal/model/deployment/BitbucketDeploymentEnvironment.java:115]: could rename this to be EnvironmentType, Type is a bit generic

**Review by EliSalter-atl — APPROVED** [review]

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/model/deployment/BitbucketDeploymentEnvironment.java
Comment (khughes-atlassian) [reply src/main/java/com/atlassian/bitbucket/jenkins/internal/model/deployment/BitbucketDeploymentEnvironment.java:115]: Yeah, I thought that since it's an inner class of `BitbucketDeploymentEnvironment` it was obvious what 'type' it was. Maybe it's better if I pull it out into it's own class

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/model/deployment/BitbucketCDCapabilities.java
Comment (khughes-atlassian) [reply src/main/java/com/atlassian/bitbucket/jenkins/internal/model/deployment/BitbucketCDCapabilities.java:20]: We usually use the variable name as the message in calls to `requireNonNull`, and the rest of the usages in this repo do it that way too.

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/model/deployment/BitbucketCDCapabilities.java
Comment (khughes-atlassian) [reply src/main/java/com/atlassian/bitbucket/jenkins/internal/model/deployment/BitbucketCDCapabilities.java:14]: > just checking that all is okay because we have

Yes, the server capability is called `deployment` and the deployment capability is called `deployments`. Maybe that's something we should change on the Bitbucket side though before it gets released so that it is more consistent

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/client/BitbucketDeploymentClientImpl.java
Comment (khughes-atlassian) [reply src/main/java/com/atlassian/bitbucket/jenkins/internal/client/BitbucketDeploymentClientImpl.java:40]: No, because `namesAndValues` is not `null`, it is an empty array. The only way it would be `null` is if I passed through a `null` array:
```java
String[] namesAndValues = null;
Headers of = Headers.of(namesAndValues);
```

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/client/BitbucketClientFactory.java
Comment (dkjellin) [inline src/main/java/com/atlassian/bitbucket/jenkins/internal/client/BitbucketClientFactory.java:44]: This is the wrong location for this I think. (arguably the build status client is in the wrong place as well). 
The client framework we have is fluent. so your client should sit on the `BitbucketRepositoryClient`. Thus to get hold of the deployment client you go
```java
clientFactory.getProjectClient(projectKey)
.getRepositoryClient(repoSlug)
.getDeploymentClient(revisionSha)
```

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/client/BitbucketDeploymentClientImpl.java
Comment (ysun-atl) [inline src/main/java/com/atlassian/bitbucket/jenkins/internal/client/BitbucketDeploymentClientImpl.java:27]: do we need to check the CD capabilities before making this request? Otherwise `getCDCapabilities()` is not used currently

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/model/deployment/BitbucketCDCapabilities.java
Comment (EliSalter-atl) [reply src/main/java/com/atlassian/bitbucket/jenkins/internal/model/deployment/BitbucketCDCapabilities.java:20]: ah was going off what we had in BitbucketCICapabilities#21, but happy to not change

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/client/BitbucketClientFactory.java
Comment (khughes-atlassian) [reply src/main/java/com/atlassian/bitbucket/jenkins/internal/client/BitbucketClientFactory.java:44]: Unfortunately doing it that way I don't have access to any BitbucketSCMRepository. If I move it to a fluent API, how do you propose we would fit in the `BitbucketSCMRepository`?

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/client/BitbucketDeploymentClientImpl.java
Comment (khughes-atlassian) [reply src/main/java/com/atlassian/bitbucket/jenkins/internal/client/BitbucketDeploymentClientImpl.java:27]: Currently, it's not used but I think that the job of this client should just be to send the data off to Bitbucket. Since this class doesn't know what should happen when Bitbucket doesn't support deployments (should it be an error, warning or just silently ignored?) I think it should be the caller's job to call the capabilities endpoint first and decide what to do based on that.

**Comment by KalleOlaviNiemitalo** [comment]
A comment on deployment tracking in general, not specifically on this pull request.

I hope it will become possible to make each Jira issue display the environments to which the associated changes have or have not been deployed. That seems rather difficult to implement, though:

- If we merge a pull request that is believed to fix an issue, and deploy the result, then Jira should show that the changes have been deployed. But if it then turns out that the fix was incomplete, and we make another pull request with more commits for the same issue, then Jira should show that not all changes have been deployed.

- If we fix an issue on the master branch and deploy those changes, then Jira should show that the fix has been deployed. If we then cherry-pick those commits to the release/1.42 branch and don't deploy that, Jira should still show that the fix has been deployed, even though it now has undeployed commits.

- If we fix an issue on the bugfix/EXAMPLE-123 branch and deploy that to a test environment, then Jira should show that the fix has been deployed. If we then implement a feature on the feature/EXAMPLE-456 branch and deploy that to the same environment, and this branch does not include the EXAMPLE-123 fix, then Jira should shown that EXAMPLE-123 was undeployed.

- If a commit message or a pull request description mentions a Jira issue key, that does not always mean the deployment status of the commit or pull request should affect the deployment status that Jira displays for the issue. For example, "Because of EXAMPLE-616, we cannot take a dependency on JAXB."

- Developers sometimes mistype Jira issue keys in commit messages. If such a mistake causes Jira to show an incorrect deployment status, then there needs to be a way to correct it.

- A system may consist of multiple components (e.g. "authentication service") that are deployed individually from different repositories, or from different commits in a monorepo. If we fix issue EXAMPLE-834 for Client X, then we want to know whether the new versions of all the related components have been deployed to Client X Production. Usually though, fixing an issue needs changes in just one or a few components, so we can consider the fix deployed as soon as the new versions of those components have been deployed; the other components do not matter for this, provided that they are compatible. So something would need to understand which components need to be deployed for EXAMPLE-834. Maybe this could be controlled with the Components field in Jira, or by adding a version-controlled file that lists the source code subdirectories that pertain to each component.

With all this complexity, I guess the most feasible solution would be in-house scripting that updates custom fields in Jira.

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/model/deployment/BitbucketDeploymentEnvironment.java
Comment (dkjellin) [inline src/main/java/com/atlassian/bitbucket/jenkins/internal/model/deployment/BitbucketDeploymentEnvironment.java:17]: `@JsonIgnoreProperties(ignoreUnknown = true)` ?

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/deployments/BitbucketDeploymentFactoryImpl.java
Comment (dkjellin) [inline src/main/java/com/atlassian/bitbucket/jenkins/internal/deployments/BitbucketDeploymentFactoryImpl.java:17]: Please remove `final` on all classes, I've only spotted it here and in `BitbucketDeploymentEnvironment` so far. I will not call it out again. final classes are a royal pain, you can't mock them, you can't extend them (which can be useful in testing or sometimes other places too). Sometimes that is desirable but those cases are far and few in between, and when they do come up we should add a comment at the top of the class explaining why it is final.

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/model/deployment/BitbucketDeploymentEnvironment.java
Comment (dkjellin) [inline src/main/java/com/atlassian/bitbucket/jenkins/internal/model/deployment/BitbucketDeploymentEnvironment.java:30]: The builder offers very little benefit for us here. It is 4 fields. Thus I think we should remove the builder completely and just telescope the constructor. We typically do not use builder in entity classes (I know we have some). 
If you very strongly want to keep it, remove this constructor and call the telescoped one below from the `build`method so we only have one constructor.

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/model/deployment/BitbucketDeploymentEnvironment.java
Comment (dkjellin) [inline src/main/java/com/atlassian/bitbucket/jenkins/internal/model/deployment/BitbucketDeploymentEnvironment.java:49]: I see you've added `equals` and `hashCode` for all entity classes, is there a reason? I'm slightly concerned that we need to add a field and forget to update `equals/hashCode` when we do. I vote for removing them unless you need them for some reason?

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/model/deployment/DeploymentState.java
Comment (dkjellin) [inline src/main/java/com/atlassian/bitbucket/jenkins/internal/model/deployment/DeploymentState.java:4]: This feels very spartan to me. There is _nothing_ here to help me decide what things mean. I know writing javadoc is hard but at the very minimum I think we should link perhaps to our public docs (which I know are not public yet).

File: src/test/java/com/atlassian/bitbucket/jenkins/internal/client/BitbucketRepositoryClientImplTest.java
Comment (dkjellin) [inline src/test/java/com/atlassian/bitbucket/jenkins/internal/client/BitbucketRepositoryClientImplTest.java:76]: I'm not convinced the `deleteWhitespace` is a good idea. I fear it may hide problems, but I think it is minor enough that it is fine.

If you static import `StandardCharsets.UTF_8` you can
```suggestion
    assertEquals(StringUtils.deleteWhitespace(requestString), StringUtils.deleteWhitespace(b.readString(UTF_8)));
```
Which I think is nicer than creating new strings in whatever locale you happen to run.

The same applies throughout this file.

**Review by ysun-atl — APPROVED** [review]

File: src/test/java/com/atlassian/bitbucket/jenkins/internal/client/BitbucketRepositoryClientImplTest.java
Comment (dkjellin) [inline src/test/java/com/atlassian/bitbucket/jenkins/internal/client/BitbucketRepositoryClientImplTest.java:95]: As you know I'm not a fan of building URLs like this. Given this is a test I think it is probably ok.
Personally I would have opted for 
```String url = new HttpUrl.Builder()
.host(BITBUCKET_BASE_URL)
.addPathSegment(..)
...
.build()
.toString()
```

I would probably also have some helper methods that return either the `HttpUrl` or the `HttpUrl.Builder` to reduce repeating code.

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/deployments/BitbucketDeploymentFactoryImpl.java
Comment (dkjellin) [inline src/main/java/com/atlassian/bitbucket/jenkins/internal/deployments/BitbucketDeploymentFactoryImpl.java:52]: I hope this is a big field. Jenkins allows around 200k for this. and it can be formatted text (html or markdown, depending on configuration). Are we sure this is what we want? Does it make sense? (not objecting just pointing things out)

File: src/test/resources/deployments/send_deployment_result.json
Comment (dkjellin) [inline src/test/resources/deployments/send_deployment_result.json:2]: could we put in the real response regardless? That would enable us to look at the response and perhaps do some validation (should we want to in the future) without first having to find the actual response. Just a thought

**Review by dkjellin — CHANGES_REQUESTED** [review]
Overall I think this looks good. `final` classes is something we should remove unless there is a strong reason to keep them? I feel so strongly about it that I requested changes to ensure it is done. The rest is minor stuff.

I am concerned with the use of `job description` for the deployment description. I know they share the same name, but I am not convinced that it is enough overlap, maybe there is, if so we just need to manage all the things that comes with using a free form text.

Lastly, I see you moved the `BuildStatusClient` kudos for that. I think we need to test an upgrade from a previous version just to ensure it does not muck around with serialised jobs. I don't think that blocks merging this to your feature branch, but we do need to test it before the feature branch is merged to master.

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/model/deployment/BitbucketDeploymentEnvironment.java
Comment (dkjellin) [reply src/main/java/com/atlassian/bitbucket/jenkins/internal/model/deployment/BitbucketDeploymentEnvironment.java:17]: You have it on the other ones. I think it is better to add it (given it is so easy to do), and that way all the new classes have them, and if we change the client in the future it is there and we don't accidentally ship a bug.

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/model/deployment/BitbucketDeploymentEnvironment.java
Comment (dkjellin) [reply src/main/java/com/atlassian/bitbucket/jenkins/internal/model/deployment/BitbucketDeploymentEnvironment.java:30]: >Is there a reason you would avoid it other than 'we typically do not use' them?

The builder is more code to maintain, sure not much but it is more. You now have two ways to construct this object.  It is unusual to have builders on entity classes. Depending on how it is done it can cause problems when serialised. If it brings value, it is worth it, but I don't see the value here. The `BitbucketDeployment` has more fields, no builder. I know that not all fields are required here, but they end up as `null` anyway so.. just pass `null` in and avoid creating a whole new object to create this one.

>since there are two fields that are optional

How would that look? I would expect that those fields are held in variables? So you could just pass them in, and if they're not set they're `null` and things just work?
if the caller has to `if (type != null) { builder.type(type)` then we've only gained more code and no benefit. 
Builders are useful, I just don't think it adds just any benefit in this case.

>You're right that I shouldn't have two constructors setting fields here. Is there a reason you'd prefer telescoping it from the builder instead of me calling this from this constructor?

It depends on the case with just 4 fields, I feel this private constructor "obscures" the one that gets called. Having a single constructor is preferable. If we need to have more than one constructor chaining them is preferred. 

Like I've said many times, I just fail to see the value this builder brings. Jenkins is **very** `null` heavy, so passing `null` for unset fields is what is done, in this plugin and other places. I feel this is just more code to maintain. 

Despite the near novel here, if you feel it is valuable I'm happy enough to leave it in.

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/model/deployment/DeploymentState.java
Comment (dkjellin) [reply src/main/java/com/atlassian/bitbucket/jenkins/internal/model/deployment/DeploymentState.java:4]: > callers can choose whichever 

I don't think this is true. We don't expose API. It is _here_ (in this plugin) that you assign meaning to these values. From discussions it may even be possible for users to not assign these and the plugin makes the appropriate decisions for itself. This I think it would be good for us to probably document a bit more around what they mean for us here in Jenkins land.

**Review by dkjellin — APPROVED** [review]
Looks good!

**Comment by khughes-atlassian** [comment]
Hi @KalleOlaviNiemitalo, thanks for your interest in this pull request. I will try to answer all of your questions here, but please know that this is a work-in-progress and interacting with a feature in Bitbucket DC that is still actively under development. The initial goal is to show deployment information on commits and pull requests.

> A comment on deployment tracking in general, not specifically on this pull request.
> 
> I hope it will become possible to make each Jira issue display the environments to which the associated changes have or have not been deployed.

The long-term goal is for this to happen so that Bitbucket DC can take over sending the deployment information to Jira, since it is better suited to finding all the commits that are part of a deployment. In the meantime, however, if you are using Jira Cloud I'd direct you to [Atlassian's Jira plugin](https://www.jenkins.io/doc/pipeline/steps/atlassian-jira-software-cloud/#jirasenddeploymentinfo-atlassian-jira-software-cloud-jenkins-integration-deployment) which supports deployments.

> That seems rather difficult to implement, though:
There are indeed a lot of scenarios with deployments that are difficult to handle. We will be focusing initially on straightforward / roll-forward scenarios and will not be automatically marking deployments as 'undeployed'. However, we aim to make the REST API public and have a range of possible deployment states to give customers maximum control.

Regards,
Kristy

File: src/main/java/com/atlassian/bitbucket/jenkins/internal/model/deployment/BitbucketDeployment.java
Comment (mhenschke-atl) [inline src/main/java/com/atlassian/bitbucket/jenkins/internal/model/deployment/BitbucketDeployment.java:29]: It's the smallest of nits, but Jenkins run numbers are always ints so the long precision isn't usable here

**Review by mhenschke-atl — APPROVED** [review]

**Comment by KalleOlaviNiemitalo** [comment]
I see in [API Changelog](https://developer.atlassian.com/server/bitbucket/reference/api-changelog/) that Bitbucket Server 7.16 now has REST endpoints for deployments. [Bitbucket Data Center and Server 7.16 release notes](https://confluence.atlassian.com/bitbucketserver/bitbucket-data-center-and-server-7-16-release-notes-1081348350.html) however do not mention it. Does this mean that the data model and REST API already exist but the user interface was postponed to a later version?

**Comment by khughes-atlassian** [comment]
> I see in [API Changelog](https://developer.atlassian.com/server/bitbucket/reference/api-changelog/) that Bitbucket Server 7.16 now has REST endpoints for deployments. [Bitbucket Data Center and Server 7.16 release notes](https://confluence.atlassian.com/bitbucketserver/bitbucket-data-center-and-server-7-16-release-notes-1081348350.html) however do not mention it. Does this mean that the data model and REST API already exist but the user interface was postponed to a later version?

@KalleOlaviNiemitalo Both the REST API and the UI exist, however since we have no integrations released in either Bamboo or Jenkins we are postponing mentioning it in the release notes until customers have a way to use the feature. Developers are able to use the API now if they wish, and we plan to retroactively update the release notes once support for Jenkins and Bamboo has been released.

**Comment by KalleOlaviNiemitalo** [comment]
> we plan to retroactively update the release notes once support for Jenkins and Bamboo has been released

The support for Jenkins was released in Bitbucket Server Integration 3.1.0 (5 November 2021), and [Bitbucket Data Center and Server 8.1 release notes](https://confluence.atlassian.com/bitbucketserver/bitbucket-data-center-and-server-8-1-release-notes-1130726463.html#BitbucketDataCenterandServer8.1releasenotes-ViewdeploymentandbuildinformationinJiraCloud) now state that Bitbucket Data Center can send deployment information to Jira Cloud. However, [Bitbucket Data Center and Server 7.16 release notes](https://confluence.atlassian.com/bitbucketserver/bitbucket-data-center-and-server-7-16-release-notes-1081348350.html) still do not mention deployments. Is that waiting for Bamboo?

**Comment by khughes-atlassian** [comment]
Thanks for pointing this out @KalleOlaviNiemitalo! We have missed retroactively updating the release notes and will get right on that.

Deployment notifications should be able to be sent to Bitbucket DC 7.16+. You'll see the deployment information on pull requests or commits that are contained in the deployment (i.e. commits that are part of the given deployment but not part of the previous deployment for the given key & environment). The Bitbucket Server release notes (when updated) will have more details. Jenkins 3.1+ and Bamboo 8.1+ also have mechanisms of sending through deployment notifications to Bitbucket DC.

**Comment by dkjellin** [comment]
Thank you for pointing this out @KalleOlaviNiemitalo, I'll get the release notes updated to be clearer.

**Pull request merged**
